crates/energy/src/graph.rs

```diff
@@ -0,0 +1,242 @@
+    fn build(&self, app: &mut App) {
+        app.add_plugins((DebugLinesPlugin::default(), DespawnEventsPlugin::<&NearbyUnits, NearbyUnits>::default()))
+            .add_systems(OnEnter(GameState::Playing), setup)
+            .add_systems(OnExit(GameState::Playing), clean_up)
+            .add_systems(Update, spawn_graph_components.after(SpawnerSet::Spawn))
+            .add_systems(
+                PreUpdate,
+                (
+                    remove_old_nodes.before(GraphSystemSet::UpdateNearby),
+                    update_nearby_recv.in_set(GraphSystemSet::UpdateNearby),
+                    update_graph.in_set(GraphSystemSet::UpdateGraph).after(GraphSystemSet::UpdateNearby),
+                    debug_lines.after(GraphSystemSet::UpdateGraph),
+                )
+                    .run_if(in_state(GameState::Playing)),
+            );
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum GraphSystemSet {
+    UpdateNearby,
+    UpdateGraph,
+}
+
+/// The power grid resource is used to store the power grid graph.
+#[derive(Resource, Debug, Clone)]
+pub(crate) struct PowerGrid {
+    /// The power grid graph.
+    graph: GraphMap<Entity, f64, Undirected>,
+}
+
+impl Default for PowerGrid {
+    fn default() -> Self {
+        Self {
+            graph: GraphMap::new(),
+        }
+    }
+}
+
+/// The energy receiver component is used to mark an entity as an energy receiver.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyReceiver;
+
+/// The energy producer component is used to mark an entity as an energy producer.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyProducer;
+
+/// The nearby component is used to store the nearby entities of an entity.
+#[derive(Debug, Clone)]
+pub enum Nearby {
+    Receiver(SmallVec<[Entity; 40]>),
+    Producer(SmallVec<[Entity; 40]>),
+}
+
+impl Nearby {
+    fn into_inner(self) -> SmallVec<[Entity; 40]> {
+        match self {
+            Nearby::Receiver(inner) => inner,
+            Nearby::Producer(inner) => inner,
+        }
+    }
+}
+
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits(SmallVec<[Nearby; 2]>, Option<Vec2>);
+
+fn setup(mut commands: Commands) {
+    commands.insert_resource(PowerGrid::default());
+}
+
+fn clean_up(mut commands: Commands) {
+    commands.remove_resource::<PowerGrid>();
+}
+
+/// This system spawns Energy Producers and Energy Receivers and nearby units.
+fn spawn_graph_components(
+    mut commands: Commands,
+    newly_spawned_units: Query<Entity, Added<Active>>,
+) {
+    for entity in newly_spawned_units.iter() {
+        commands
+            .entity(entity)
+            .insert((EnergyReceiver, NearbyUnits::default()));
+    }
+}
+
+fn update_nearby_recv(
+    spacial_index_producer: SpatialQuery<Entity, With<EnergyProducer>>,
+    spacial_index_receiver: SpatialQuery<Entity, With<EnergyReceiver>>,
+    mut nearby_units: Query<(Entity, &mut NearbyUnits, &Transform)>,
```

Comment:
I would add `, Changed<Transform>` because even the `if transform.translation.to_flat().distance_squared(last_pos) < 0.5 {` is not free.